File: v2/simulate.js

const kerf = 5;
const binWidth = 2440;
const binHeight = 1220;

class SingleBin {
    constructor(width, height, kerf) {
        this.width = width;
        this.height = height;
        this.kerf = kerf;
        this.freeRects = [{ x: 0, y: 0, width, height }];
        this.placed = [];
        this.cuttingCount = 0;
    }

    pack(items) {
        const unplaced = [];
        for (const item of items) {
            const result = this.insert(item.width, item.height, item.rotatable);
            if (result) {
                this.placed.push({ ...item, x: result.x, y: result.y, width: result.width, height: result.height });
            } else {
                unplaced.push(item);
            }
        }
        return { placed: this.placed, unplaced, cuttingCount: this.cuttingCount };
    }

    insert(width, height, rotatable) {
        let bestRect = null;
        let bestRectIndex = -1;
        let bestShortSideFit = Infinity;
        let bestRotated = false;

        for (let i = 0; i < this.freeRects.length; i++) {
            const rect = this.freeRects[i];
            if (rect.width >= width + this.kerf && rect.height >= height + this.kerf) {
                const shortSide = Math.min(rect.width - width, rect.height - height);
                if (shortSide < bestShortSideFit) {
                    bestRect = rect; bestRectIndex = i; bestShortSideFit = shortSide; bestRotated = false;
                }
            }
            if (rotatable && rect.width >= height + this.kerf && rect.height >= width + this.kerf) {
                const shortSide = Math.min(rect.width - height, rect.height - width);
                if (shortSide < bestShortSideFit) {
                    bestRect = rect; bestRectIndex = i; bestShortSideFit = shortSide; bestRotated = true;
                }
            }
        }

        if (!bestRect) return null;
        const pW = bestRotated ? height : width;
        const pH = bestRotated ? width : height;
        this.splitFreeRect(bestRectIndex, pW + this.kerf, pH + this.kerf);
        return { x: bestRect.x, y: bestRect.y, width: pW, height: pH, rotated: bestRotated };
    }

    splitFreeRect(index, uW, uH) {
        const rect = this.freeRects[index];
        const w = rect.width - uW;
        const h = rect.height - uH;
        if (w <= h) {
            if (w > 0) { this.freeRects.push({ x: rect.x + uW, y: rect.y, width: w, height: uH }); this.cuttingCount++; }
            if (h > 0) { this.freeRects.push({ x: rect.x, y: rect.y + uH, width: rect.width, height: h }); this.cuttingCount++; }
        } else {
            if (h > 0) { this.freeRects.push({ x: rect.x, y: rect.y + uH, width: uW, height: h }); this.cuttingCount++; }
            if (w > 0) { this.freeRects.push({ x: rect.x + uW, y: rect.y, width: w, height: rect.height }); this.cuttingCount++; }
        }
        this.freeRects.splice(index, 1);
    }
}

const items = [];
for (let i = 0; i < 6; i++) {
    items.push({ width: 1000, height: 600, rotatable: true });
}

let totalCuts = 0;
let remaining = [...items];
let binIdx = 1;
while (remaining.length > 0) {
    const bin = new SingleBin(binWidth, binHeight, kerf);
    const result = bin.pack(remaining);
    if (result.placed.length === 0) break;
    console.log(`Bin ${binIdx}: Placed ${result.placed.length}, Cuts: ${result.cuttingCount}`);
    totalCuts += result.cuttingCount;
    remaining = result.unplaced;
    binIdx++;
}
console.log(`Total Cuts: ${totalCuts}`);

File: v2/simulate.py

kerf = 5
bin_width = 2440
bin_height = 1220

class SingleBin:
    def __init__(self, width, height, kerf):
        self.width = width
        self.height = height
        self.kerf = kerf
        self.free_rects = [{"x": 0, "y": 0, "width": width, "height": height}]
        self.placed = []
        self.cut_lines_x = set()
        self.cut_lines_y = set()

    def pack(self, items):
        unplaced = []
        for item in items:
            result = self.insert(item["width"], item["height"], item.get("rotatable", True))
            if result:
                self.placed.append({**item, **result})
            else:
                unplaced.append(item)
        return {
            "placed": self.placed, 
            "unplaced": unplaced, 
            "cutting_count": len(self.cut_lines_x) + len(self.cut_lines_y)
        }

    def insert(self, width, height, rotatable):
        best_rect = None
        best_rect_index = -1
        best_short_side_fit = float('inf')
        best_rotated = False

        for i, rect in enumerate(self.free_rects):
            # Normal
            if rect["width"] >= width + self.kerf and rect["height"] >= height + self.kerf:
                short_side = min(rect["width"] - width, rect["height"] - height)
                if short_side < best_short_side_fit:
                    best_rect = rect
                    best_rect_index = i
                    best_short_side_fit = short_side
                    best_rotated = False
            
            # Rotated
            if rotatable and rect["width"] >= height + self.kerf and rect["height"] >= width + self.kerf:
                short_side = min(rect["width"] - height, rect["height"] - width)
                if short_side < best_short_side_fit:
                    best_rect = rect
                    best_rect_index = i
                    best_short_side_fit = short_side
                    best_rotated = True

        if not best_rect:
            return None

        pW = height if best_rotated else width
        pH = width if best_rotated else height
        self.split_free_rect(best_rect_index, pW + self.kerf, pH + self.kerf)
        return {"x": best_rect["x"], "y": best_rect["y"], "width": pW, "height": pH, "rotated": best_rotated}

    def split_free_rect(self, index, uW, uH):
        rect = self.free_rects[index]
        w = rect["width"] - uW
        h = rect["height"] - uH

        if w <= h:
            if w > 0:
                self.free_rects.append({"x": rect["x"] + uW, "y": rect["y"], "width": w, "height": uH})
                self.cut_lines_x.add(rect["x"] + uW)
            if h > 0:
                self.free_rects.append({"x": rect["x"], "y": rect["y"] + uH, "width": rect["width"], "height": h})
                self.cut_lines_y.add(rect["y"] + uH)
        else:
            if h > 0:
                self.free_rects.append({"x": rect["x"], "y": rect["y"] + uH, "width": uW, "height": h})
                self.cut_lines_y.add(rect["y"] + uH)
            if w > 0:
                self.free_rects.append({"x": rect["x"] + uW, "y": rect["y"], "width": w, "height": rect["height"]})
                self.cut_lines_x.add(rect["x"] + uW)
        self.free_rects.pop(index)

items = [{"width": 1000, "height": 600, "rotatable": False} for _ in range(6)]
total_cuts = 0
remaining = items
bin_idx = 1

while len(remaining) > 0:
    bin_obj = SingleBin(bin_width, bin_height, kerf)
    result = bin_obj.pack(remaining)
    if not result["placed"]:
        break
    print(f"Bin {bin_idx}: Placed {len(result['placed'])}, Cuts: {result['cutting_count']}")
    total_cuts += result["cutting_count"]
    remaining = result["unplaced"]
    bin_idx += 1

print(f"Total Cuts: {total_cuts}")
print(f"Total Cost (18mm): {total_cuts * 1500} KRW")
